src/adapters/pnp/Watcher.c

```diff
@@ -0,0 +1,249 @@
+        
+        if ((MPI_OK == mpiResult) && (NULL != payload) && (0 < payloadSizeBytes))
+        {
+            if (*hash != (payloadHash = HashString(payload)))
+            {
+                if (SavePayloadToFile(fileName, payload, payloadSizeBytes, GetLog()))
+                {
+                    RestrictFileAccessToCurrentAccountOnly(fileName);
+                    *hash = payloadHash;
+                }
+            }
+        }
+        
+        CallMpiFree(payload);
+    }
+}
+
+static void ProcessDesiredConfigurationFromFile(const char* fileName, size_t* hash, void* log)
+{
+    size_t payloadHash = 0;
+    int payloadSizeBytes = 0;
+    char* payload = NULL; 
+    bool platformAlreadyRunning = true;
+    int mpiResult = MPI_OK;
+
+    if (fileName && hash)
+    {
+        RestrictFileAccessToCurrentAccountOnly(fileName);
+
+        payload = LoadStringFromFile(fileName, false, GetLog());
+        if (payload && (0 != (payloadSizeBytes = strlen(payload))))
+        {
+            // Do not call MpiSetDesired unless this desired configuration is different from previous
+            if (*hash != (payloadHash = HashString(payload)))
+            {
+                OsConfigLogInfo(log, "Watcher processing DC payload from %s", fileName);
+
+                mpiResult = CallMpiSetDesired((MPI_JSON_STRING)payload, payloadSizeBytes, GetLog());
+                if ((MPI_OK != mpiResult) && RefreshMpiClientSession(&platformAlreadyRunning) && (false == platformAlreadyRunning))
+                {
+                    mpiResult = CallMpiSetDesired((MPI_JSON_STRING)payload, payloadSizeBytes, GetLog());
+                }
+            
+                if (MPI_OK == mpiResult)
+                {
+                    *hash = payloadHash;
+                }
+            }
+        }
+        FREE_MEMORY(payload);
+    }
+}
+
+static int InitializeGitClone(const char* gitRepositoryUrl, const char* gitClonePath, void* log)
+{
+    const char* g_gitCloneCleanUpTemplate = "rm -r %s";
+    const char* g_gitCloneTemplate = "git clone -q %s %s";
+    const char* g_gitConfigTemplate = "git config --global --add safe.directory %s";
+    
+    char* cleanUpCommand = NULL;
+    char* configCommand = NULL;
+    char* cloneCommand = NULL;
+    int error = 0;
+
+    if ((NULL == gitRepositoryUrl) || (NULL == gitClonePath))
+    {
+        OsConfigLogError(log, "InitializeGitClone: invalid arguments");
+        return EINVAL;
+    }
+
+    // Do not log gitRepositoryUrl as it may contain Git account credentials
+
+    cleanUpCommand = FormatAllocateString(g_gitCloneCleanUpTemplate, gitClonePath);
+    cloneCommand = FormatAllocateString(g_gitCloneTemplate, gitRepositoryUrl, gitClonePath);
+    configCommand = FormatAllocateString(g_gitConfigTemplate, gitClonePath);
+    
+    ExecuteCommand(NULL, cleanUpCommand, false, false, 0, 0, NULL, NULL, log);
+
+    if (0 != (error = ExecuteCommand(NULL, cloneCommand, false, false, 0, 0, NULL, NULL, GetLog())))

```

Comment:
>ExecuteCommand [](http://example.com/codeflow?start=22&length=14)

Is it OK to fail with the below message if the *git* commands are not installed on the system? Or should there be an explicit check for that first?

Reply:
Good suggestion. "Git does not appear installed on the local system". We can have this as a prerequisite in documentation (to be added later) or also a check in code. This can be done in a separate PR, let's continue to talk about it offline.